0_fund_analysis_functions_v2.R

```diff
@@ -0,0 +1,1210 @@
+  # pacta_company_debt_results <- read_rds(paste0(project_location, "40_Results/Bonds_results_company.rds"))
+  # pacta_company_equity_results <- read_rds(paste0(project_location, "40_Results/Equity_results_company.rds"))
+  # 
+  # pacta_company_debt_results$asset_type <- "Bonds"
+  # pacta_company_equity_results$asset_type <- "Equity"
+  # 
+  # pacta_company_results <- pacta_company_debt_results %>% 
+  #   bind_rows(pacta_company_equity_results)
+  # 
+  # pacta_company_results <- pacta_company_results %>% 
+  #   filter((ald_sector == "Power" & scenario_geography == "GlobalAggregate") | (ald_sector != "Power" & scenario_geography == "Global")
+  #   )
+  # 
+  # pacta_company_results <- pacta_company_results %>% 
+  #   filter(investor_name != portfolio_name | portfolio_name == "Meta Portfolio")
+  # 
+  # pacta_company_results <- pacta_company_results %>% 
+  #   clean_names(case = "snake") %>% 
+  #   mutate_at(
+  #     .vars = c("ald_sector", "technology", "asset_type"),
+  #     .funs = tolower
+  #   )
+  # 
+  # input_pacta_company_results <<- pacta_company_results
+}
+
+
+check_group_share_parameters <- function(
+  .data, 
+  id_cols,
+  numerator_group, 
+  denominator_group,
+  values_from,
+  values_to
+) {
+  character_parameters <- c(numerator_group, denominator_group, values_from)
+  character_null_parameters <- c(id_cols, values_to)
+  # check parameters 
+  lapply(character_parameters, function(x) {stopifnot(is.character(x) & !is.null(x))})
+  lapply(character_null_parameters, function(x) {stopifnot(is.character(x) | is.null(x))})
+  # check names 
+  crucial_names <- c(values_from, numerator_group, denominator_group, id_cols)
+  r2dii.utils::check_crucial_names(.data, crucial_names)
+  # check values
+  assertr::verify(.data, class(.data[[values_from]]) == "numeric")
+  # return 
+  return(.data)
+}
+
+summarise_group_share <- function(
+  .data,
+  id_cols = NULL,
+  numerator_group = "asset_type", 
+  denominator_group = "portfolio_name",
+  values_from = "value_usd",
+  values_to = NULL,
+  ...
+) {
+  # ungroup data 
+  .data <- .data %>% 
+    ungroup()
+  # check parameters
+  .data <- .data %>% 
+    check_group_share_parameters(
+      id_cols = id_cols,
+      numerator_group = numerator_group, 
+      denominator_group = denominator_group,
+      values_from = values_from,
+      values_to = values_to
+    )
+  # define group_vars 
+  group_vars <- unique(c(id_cols, denominator_group))
+  # calculate exposure 
+  .data <- .data %>%
+    dplyr::group_by(!!!rlang::syms(c(group_vars, numerator_group))) %>% 
+    dplyr::summarise(group_share = sum(.data[[values_from]], ...)) %>% 
+    group_by(!!!rlang::syms(group_vars)) %>% 
+    dplyr::mutate(group_share = .data$group_share / sum(.data$group_share, ...)) %>% 
+    dplyr::ungroup()
+  # rename output
+  if (!is.null(values_to)) {
+    .data %>% 
+      dplyr::rename("{values_to}" := .data$group_share)
+  } else if (is.null(values_to)) {
+    return(.data)
```

Comment:
Reserve `return()` for returning early.
See 3.3 at https://style.tidyverse.org/functions.html